aws/sdk/examples/telephone-game/src/main.rs

```diff
@@ -0,0 +1,439 @@
+
+        // Set up a temp directory to store audio files
+        let tmp_dir = TempDir::new("telephone-game").expect("couldn't create temp dir");
+        let tmp_file_path = tmp_dir.path().join("polly.mp3");
+
+        // Set up the ability to play audio
+        let (_stream, stream_handle) =
+            OutputStream::try_default().expect("Couldn't get handle to default audio output");
+        let sink = Sink::try_new(&stream_handle).unwrap();
+
+        loop {
+            info!(r#"Say something and I'll echo it or say "quit" to quit"#);
+            // Collect user input
+            let mut input_lines = BufReader::new(tokio::io::stdin()).lines();
+
+            // Reading a line from stdin will include the trailing newline
+            // We don't want that so we trim it
+            line = input_lines
+                .next_line()
+                .await
+                .expect("encountered IO error when attempting to read next line")
+                .expect("stream is closed")
+                .trim()
+                .to_owned();
+
+            // If the user types "quit" then the game will end
+            if line == "quit" {
+                break;
+            }
+
+            // Start synthesizing speech and wait for it to finish
+            let res = polly_client
+                .synthesize_speech()
+                .text(&line)
+                .voice_id(VoiceId::Joanna)
+                .output_format(OutputFormat::Mp3)
+                .send()
+                .await;
+
+            match res {
+                Ok(res) => {
+                    info!(r#"You said: "{}""#, &line);
+                    info!("Playing Polly's response...");
+
+                    // Collect the ByteStream returned by the synthesize_speech call
+                    let byte_stream = res
+                        .audio_stream
+                        .collect()
+                        .await
+                        .expect("audio stream ended prematurely");
+
+                    // Create a file to store the audio
+                    let mut tmp_file = tokio::fs::File::create(&tmp_file_path)
+                        .await
+                        .expect("couldn't create temp file");
+                    // Write the ByteStream to the file
+                    tmp_file
+                        .write_all(&byte_stream.into_bytes())
+                        .await
+                        .expect("couldn't write to temp file");
+                    // Flush the write operation to ensure it finishes before we continue
+                    tmp_file.flush().await.expect("couldn't flush");
+
+                    // Open the audio file with regular blocking IO File
+                    // rodio's Decoder requires stdlib Files
+                    let file = std::fs::File::open(&tmp_file_path).unwrap();
+                    let source =
+                        Decoder::new(std::io::BufReader::new(file)).expect("couldn't decode audio");
+
```

Comment:
I don't know if we _should_, but you could write the response data from the ByteStream as it came off the wire to a sync::mpsc::receiver (which implements `Iter<T>`). Then use iter-read to create something you can write to rodio.

Probably not worth it though

Reply:
I refactored this to use less async in general